### ISSUE TITLE
OnlyOffice 4.0

### ISSUE THREAD

**Issue by sneer2016** [cetra3/onlyoffice-alfresco#12]
Hi,
Ascensio System SIA has been released OnlyOffice 4.0 and I found some bugs:
1. When I open document in onlyoffice, I have a warning: The document could not be saved. Please check connection settings or contact your administrator. When you click the "OK" button, you will be prompted to download document. If i closed this message, then document is opened and i can to edit it and save.
2. Onlyoffice shows all users as anonymous. Username from Alfresco is not shows in only office.
3. Language and regional standards always is english. Can you add manual language change feature in plugin?

Thanks!

![2016-07-13 08-47-35](https://cloud.githubusercontent.com/assets/20431401/16791250/de998476-48de-11e6-8f65-316d7a291acb.png)

**Comment by cetra3** [comment]
For `1.`: This may mean that the OnlyOffice can't download the file from Alfresco.  

OnlyOffice will make the connection to Alfresco on behalf of the client, so OnlyOffice needs to be able to talk to Alfresco.  In order for OnlyOffice to do this, Alfresco needs to generate what it thinks the external URL is.  Make sure that the following properties are set correctly in `alfresco-global.properties`

```
alfresco.protocol=http
alfresco.host=alfresco.yourcompany.local
alfresco.port=8080
alfresco.context=alfresco
```

Alfresco will provide a download URL for the document which uses [UrlUtil](https://github.com/Alfresco/community-edition/blob/master/projects/repository/source/java/org/alfresco/util/UrlUtil.java) which in turn uses [SysAdminParams](https://github.com/Alfresco/community-edition/blob/master/projects/repository/source/java/org/alfresco/repo/admin/SysAdminParamsImpl.java) which in turn is driven by the above properties.  The [URL](https://github.com/cetra3/onlyoffice-alfresco/blob/master/repo/src/main/java/com/parashift/onlyoffice/Prepare.java) that is created is:

```
UrlUtil.getAlfrescoUrl(sysAdminParams) + "/s/api/node/content/workspace/SpacesStore/" + nodeRef.getId() + "?alf_ticket=" + authenticationService.getCurrentTicket();
```

If you view the source of the page, you should see the callback URL set correctly in editorConfig:

```
            editorConfig: {
                mode: "edit",
                canAutosave: false,
                callbackUrl: "http://alfresco.yourcompany.local:8080/alfresco/s/parashift/onlyoffice/callback?nodeRef=workspace://SpacesStore/538890f2-fa6b-4f9c-8c0f-6cd0d46b50ff&alf_ticket=TICKET_cf98656437c38b6c8bbc23babd90b15ea456918e",
                user: {
                  id: "admin",
                  name: "Administrator"
                }
            },
```

As for `2.`, it appears as though the `name` object in the user is deprecated and replaced with `firstname` and `lastname`.  Will need to submit a change to address this.

For `3.` would need to expose this as a configuration variable so the `lang` param is set correctly.  Will need to submit a change to address this.

**Comment by sneer2016** [comment]
For 1.: In sources of the page i see this:
            editorConfig: {
                mode: "edit",
                canAutosave: false,
                callbackUrl: "http://ecm.chofoms.ru/alfresco/s/parashift/onlyoffice/callback?nodeRef=workspace://SpacesStore/0b988586-439d-46da-8501-b4b840a03698&alf_ticket=TICKET_e8812d958891ababfdd8e735f00314f2c9888f2b",
                user: {
                  id: "admin",
                  name: "Administrator"
                }
            },
Before upgrade OnlyOffice from 3.0 version to 4.0 version i never see a warning message "The document could not be saved". In alfresco-global.properties i found this:

alfresco.context=alfresco
alfresco.host=ecm.chofoms.ru
alfresco.port=80
alfresco.protocol=http

So...i can open document, i can edit it and correctly save it. May be in OnlyOffice 4.0 is require  something else except CallBackUrl? Could you check a work of installation Alfresco 5.0+OnlyOffice 4.0 ?

 For 2-3:
In /opt/alfresco-5.0.c/tomcat/webapps/share/WEB-INF/lib/onlyoffice-alfresco-share-1.0.3.jar/alfresco/site-webscripts/com/parashift/onlyoffice-edit.get.html.flt:

```
        editorConfig: {
            mode: "edit",
            lang: "ru",
            canAutosave: false,
            callbackUrl: "${callbackUrl}",
            user: {
              id: "${userId}",
              firstname: "${userId}"
            }
        },
```

And now interface and regional standard sets for russian and i see user's login instead "anonymous".

**Comment by cetra3** [comment]
@sneer2016,

After hearing about this issue I did do a test with the following versions:
- Alfresco `5.1`
- OnlyOffice `4`

You might see if there is something funny going on within the logs if you enable debug in alfresco/extension/custom-log4j.properties:

```
log4j.logger.com.parashift=debug
```

What you should see is something like this:

```
2016-07-20 07:39:16,497 DEBUG [com.parashift.onlyoffice.Prepare] [http-bio-8080-exec-13] Sending JSON prepare object
2016-07-20 07:39:16,498 DEBUG [com.parashift.onlyoffice.Prepare] [http-bio-8080-exec-13] {
   "callbackUrl": "http://onlyofficealf:8080/alfresco/s/parashift/onlyoffice/callback?nodeRef=workspace://SpacesStore/b058be60-f1cd-46f1-b1e2-bca100cc254c&alf_ticket=TICKET_45249903c442cbe9e3dd9aa04e89dd831ba0611d",
   "docTitle": "Alfresco Metadata.docx",
   "docUrl": "http://onlyofficealf:8080/alfresco/s/api/node/content/workspace/SpacesStore/b058be60-f1cd-46f1-b1e2-bca100cc254c?alf_ticket=TICKET_45249903c442cbe9e3dd9aa04e89dd831ba0611d",
   "key": "b058be60-f1cd-46f1-b1e2-bca100cc254c_20160714013039",
   "onlyofficeUrl": "http://onlyoffice/"
}
2016-07-20 07:39:17,936 DEBUG [com.parashift.onlyoffice.CallBack] [http-bio-8080-exec-9] Received JSON Callback
2016-07-20 07:39:17,937 DEBUG [com.parashift.onlyoffice.CallBack] [http-bio-8080-exec-9] {
   "actions": [{
      "type": 1,
      "userid": "admin"
   }],
   "key": "b058be60-f1cd-46f1-b1e2-bca100cc254c_20160714013039",
   "status": 1,
   "users": ["admin"]
}
2016-07-20 07:39:17,937 DEBUG [com.parashift.onlyoffice.CallBack] [http-bio-8080-exec-9] Document open for editing, locking document
2016-07-20 07:41:11,751 DEBUG [com.parashift.onlyoffice.CallBack] [http-bio-8080-exec-22] Received JSON Callback
2016-07-20 07:41:11,752 DEBUG [com.parashift.onlyoffice.CallBack] [http-bio-8080-exec-22] {
   "actions": [{
      "type": 0,
      "userid": "admin"
   }],
   "changeshistory": "[{\"userid\":\"admin\",\"username\":\"Anonymous\",\"date\":\"2016-07-20 07:39:48\"}]",
   "changesurl": "http://onlyoffice/cache/files/b058be60-f1cd-46f1-b1e2-bca100cc254c_20160714013039_1655/changes.zip/changes.zip?md5=8b6CEXVTFYK_cM2spJsLrw==&expires=1469605272&disposition=attachment&ooname=output.zip",
   "key": "b058be60-f1cd-46f1-b1e2-bca100cc254c_20160714013039",
   "status": 2,
   "url": "http://onlyoffice/cache/files/b058be60-f1cd-46f1-b1e2-bca100cc254c_20160714013039_1655/output.docx/output.docx?md5=jd2yHmICwTxHdORU3nsFXA==&expires=1469605272&disposition=attachment&ooname=output.docx",
   "users": ["admin"]
}
2016-07-20 07:41:11,752 DEBUG [com.parashift.onlyoffice.CallBack] [http-bio-8080-exec-22] Document Updated, changing content
2016-07-20 07:41:11,761 DEBUG [com.parashift.onlyoffice.CallBack] [http-bio-8080-exec-22] Retrieving URL:http://onlyoffice/cache/files/b058be60-f1cd-46f1-b1e2-bca100cc254c_20160714013039_1655/output.docx/output.docx?md5=jd2yHmICwTxHdORU3nsFXA==&expires=1469605272&disposition=attachment&ooname=output.docx
```

This shows:
- The prepare URL to generate the editorConfig json
- The JSON Callback for when the user (admin) is open for editing, which is status `1`
- The JSON Callback for when the user is finished editing the document, which is status `2`
- Alfresco downloading the document from onlyoffice

For 2-3 I'd like to expose the lang as a global property so you can configure it, perhaps a PR from yourself?

**Comment by cetra3** [comment]
@sneer2016,

This is resolved in version `1.1.0`.  Please see the changelog in the readme.

Thanks
